x/examples/outline-cli/dns_linux.go

```diff
@@ -27,40 +28,77 @@ const (
 	resolvConfHeadBackupFile = "/etc/resolv.head.outlinecli.backup"
 )
 
+// systemDNSBackup info about systemdns backup
+type systemDNSBackup struct {
+	original, backup string
+}
+
+var systemDNSBackups = make([]systemDNSBackup, 0, 2)
+
 func setSystemDNSServer(serverHost string) error {
 	setting := []byte(`# Outline CLI DNS Setting
 # The original file has been renamed as resolv[.head].outlinecli.backup
 nameserver ` + serverHost + "\n")
 
-	if err := backupAndWriteFile(resolvConfFile, resolvConfBackupFile, setting); err != nil {
-		return err
+	err := backupAndWriteFile(resolvConfFile, resolvConfBackupFile, setting)
+	if err != nil {
+		return fmt.Errorf("dns backup: %w", err)
 	}
-	return backupAndWriteFile(resolvConfHeadFile, resolvConfHeadBackupFile, setting)
-}
 
-func restoreSystemDNSServer() {
-	restoreFileIfExists(resolvConfBackupFile, resolvConfFile)
-	restoreFileIfExists(resolvConfHeadBackupFile, resolvConfHeadFile)
+	err = backupAndWriteFile(resolvConfHeadFile, resolvConfHeadBackupFile, setting)
+	if err != nil {
+		return fmt.Errorf("dns head backup: %w", err)
+	}
+
+	return nil
 }
 
 func backupAndWriteFile(original, backup string, data []byte) error {
-	if err := os.Rename(original, backup); err != nil {
-		return fmt.Errorf("failed to backup DNS config file '%s' to '%s': %w", original, backup, err)
+	if _, err := os.Stat(original); err == nil {
+		// original file exist - move it into backup
+		if err := os.Rename(original, backup); err != nil {
+			return fmt.Errorf("failed to backup DNS config file '%s' to '%s': %w", original, backup, err)
+		}
+	} else if !errors.Is(err, os.ErrNotExist) { // if not exist - it's ok, just write to it
+		return errors.New("check original file exists")
 	}
-	if err := os.WriteFile(original, data, 0644); err != nil {
+
+	// save data to original
+	if err := os.WriteFile(original, data, 0o644); err != nil {
 		return fmt.Errorf("failed to write DNS config file '%s': %w", original, err)
 	}
+
+	systemDNSBackups = append(systemDNSBackups, systemDNSBackup{
+		original: original,
+		backup:   backup,
+	})
+
```

Comment:
I'd suggest to put these lines before `os.WriteFile`. Because we have already backed up the original DNS config, even though we failed to write to the new file, we still need to copy the backup file back to original when we try to reset the DNS config.

Reply:
done